README
Gtk2/Net/LDAP/Widgets version 2.0.0
==================================

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  Gtk2
  Net::LDAP

COPYRIGHT AND LICENCE

Copyright (C) 2005,2008 Aleksander Adamowski

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself. 



Changes
Revision history for Perl extension Gtk2::Net::LDAP::Widgets

2.0.0  Tue, 08 Apr 2008 13:18:12 +0200 aleksander.adamowski@gmail.com
        - initial public release


Widgets.pm
package Gtk2::Net::LDAP::Widgets;

our $VERSION = "2.0.0";
sub Version { $VERSION; }

require 5.005;

use strict;
use Gtk2::Net::LDAP::Widgets::LdapEntrySelector;
use Gtk2::Net::LDAP::Widgets::LdapEntryView;
use Gtk2::Net::LDAP::Widgets::LdapTreeSelector;
use Gtk2::Net::LDAP::Widgets::LdapTreeView;

1;

__END__

=head1 NAME

Gtk2::Net::LDAP::Widgets - LDAP-related widget library for Gtk2

=head1 SYNOPSIS

    use Gtk2::Net::LDAP::Widgets;

    my $entryPopup = Gtk2::Net::LDAP::Widgets::LdapEntrySelector->new ($parent_window,
      $ldap_source,
      'dc=example,dc=com',
      'objectClass=inetorgperson',
      'init_interactive_filter' => 'smith',
      'single_selection' => 1,
      'interactive_filter_type' => 'simple'
    );
    $entryPopup->signal_connect (response => sub {
      my ($popup, $response) = @_;
      if($response =~ 'accept') {
        print "Selected entry DN: ".$entryPopup->get_dn;
      } else {
				print "No existing entry selected.\n";
      }
      $_[0]->destroy;
      });
    $entryPopup->show_all;

=head1 DESCRIPTION

This is an object oriented collection of LDAP-related Gtk2 widgets.

Featured classes currently include windows and views that allow the user to 
select or pick entries that result from an LDAP search.

See also:

L<Gtk2::Net::LDAP::Widgets::LdapEntrySelector>
L<Gtk2::Net::LDAP::Widgets::LdapTreeSelector>
L<Gtk2::Net::LDAP::Widgets::LdapTreeView>

=head1 AUTHORS

Original author: Aleksander Adamowski <aleksander.adamowski@gmail.com>

=head1 COPYRIGHT AND LICENSE

Copyright 2005,2008 by Aleksander Adamowski

This library is free software; you can redistribute it and/or modify it under
the same terms as Perl itself. 

=cut



Widgets/Util.pm
package Gtk2::Net::LDAP::Widgets::Util;
#---[ pod head ]---{{{

=head1 NAME

Gtk2::Net::LDAP::Widgets::Util - helper functions

=head1 SYNOPSIS

    This module contains various helper functions and isn't meant to be used 
    directly. Read the source in case of any needs to do that.

=cut

#---}}}

require Exporter;
our @ISA     = qw(Exporter);
our @EXPORT  = qw(filter_trim_outer_parens);
our $VERSION = "1.9.9";

use strict 'vars';

#---[ sub filter_trim_outer_parens ]---{{{

=head2 filter_trim_outer_parens

=over 4

=item filter_trim_outer_parens ( filter )

Trims superfluous outside parentheses from an LDAP filter, e.g. ((uid=olo)) 
will be changed to uid=olo

C<filter> string representation of LDAP filter to trim parentheses from.

=back

=cut
sub filter_trim_outer_parens {
  my $filter = shift;
  while ($filter =~ /^\(.*\)$/) {
    $filter =~ s/^\((.*)\)$/$1/;
  }
  return $filter;
}
#---}}}

1;
__END__

#---[ pod end ]---{{{

=head1 SEE ALSO

L<Gtk2::Net::LDAP::Widgets>
L<Gtk2>
L<Net::LDAP>

=head1 AUTHOR

Aleksander Adamowski, E<lt>aleksander.adamowski@gmail.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright 2005,2008 by Aleksander Adamowski

This library is free software; you can redistribute it and/or modify it under
the same terms as Perl itself. 

=cut

#---}}}



Widgets/LdapTreeView.pm
package Gtk2::Net::LDAP::Widgets::LdapTreeView;
#---[ pod head ]---{{{

=head1 NAME

Gtk2::Net::LDAP::Widgets::LdapEntryView - LDAP entry viewport

=head1 SYNOPSIS

    use Gtk2::Net::LDAP::Widgets;

  $treeView = Gtk2::Net::LDAP::Widgets::LdapTreeView->new($ldap_source, 'ou=OrgStructure,dc=example,dc=com', 'objectClass=top');
	# expand entries two tree levels below:
  $treeView->expand_row(Gtk2::TreePath->new_from_string('0'), 0);
  $treeView->expand_row(Gtk2::TreePath->new_from_string('0:0'), 0);
	# ... later ...
	print join(", ", $treeView->get_dn);
=head1 ABSTRACT

Gtk2::Net::LDAP::Widgets::LdapEntryView is a child class to L<Gtk2::TreeView> 
and is used to create a Gtk2 component which lets the user select LDAP entry/entries
displayed in a tree-like structure.

Note: there might be problems with displaying the tree when an interactive 
filter is set since there may be problems building the tree if all ancestors 
for an entry aren't included in the search result.

So it's advised to carefully control the filters that are fed to this component.

=cut

#---}}}
use utf8;
use strict;
use vars qw(@ISA $VERSION);

use Net::LDAP;
use Net::LDAP::Util;
use Gtk2 -init;
use Data::Dumper;
use Gtk2::Net::LDAP::Widgets::DistinguishedName;
use Gtk2::Net::LDAP::Widgets::Util;

@ISA = qw(Gtk2::TreeView);

our $VERSION = "1.9.9";

our $rdn_column = 0;
our $bool_column = 1;
our $dn_column = 2;


use overload
q{""} => 'to_string';

#---[ sub new ]---{{{

=head1 CONSTRUCTOR

=over 4

=item new ( ldap_source, base_dn, static_filter, interative_filter, single_selection)

Creates a new Gtk2::Net::LDAP::Widgets::LdapTreeView object.

C<ldap_source> the L<Net::LDAP> object which is an active connection to an LDAP server

C<base_dn> the base DN of LDAP search operations

C<static_filter> the static filter that will be logically AND-ed with all filters executed by this selector

C<interactive_filter> the additional filter that usually comes from filter box components

C<single_selection> whether to use single selection mode (otherwise multiple selection is posible)

=back

=cut
sub new {
  my $class = shift;
  my $self = $class->SUPER::new;
  $self->{ldap_source} = shift;
  $self->{base_dn} = shift;
  $self->{static_filter} = shift;
  defined($self->{static_filter}) or $self->{static_filter} = '';
  $self->{interactive_filter} = shift;
  defined($self->{interactive_filter}) or $self->{interactive_filter} = '';
  $self->{single_selection} = shift;

  $self->{selectedDN} = undef;



  bless $self, $class;
  $self->set_rules_hint (1);
  $self->get_selection->set_mode ('multiple');
  ###
  # 1st column
  my $renderer = Gtk2::CellRendererText->new;
  my $col_offset = $self->insert_column_with_attributes
    (-1, "Entry", $renderer,
     text => 0);
  my $column = $self->get_column ($col_offset - 1);
  $column->set_clickable (1);

  ###
  # 2nd column
  $renderer = Gtk2::CellRendererToggle->new;
  $renderer->set (xalign => 0.0);
  $renderer->set_data (column => 1);
  if ($self->{single_selection}) {
    $renderer->set_radio (1);
  }
  $renderer->signal_connect (toggled => \&LdapTreeView_item_toggled, $self);
  $col_offset = $self->insert_column_with_attributes
    (-1, "Y/N", $renderer,
     active => 1
    );
  $column = $self->get_column ($col_offset - 1);
  $column->set_clickable (1);
  #$column->set_sizing ('fixed');
  #$column->set_fixed_width (50);

  $self->refresh_model;
  bless $self, $class;
}
#---}}}

#---[ sub refresh_model ]---{{{
=head2 refresh_model

Refresh the data model - re-execute the search with the current filters

=cut
sub refresh_model {
  my $self = shift;
  my $static_filter = $self->{static_filter};
  my $interactive_filter = $self->{interactive_filter};
  # Remove superfluous pairs of parentheses:
  $interactive_filter = filter_trim_outer_parens($interactive_filter);
  $static_filter = filter_trim_outer_parens($static_filter);

  my $compositeFilter;
  if (length($interactive_filter) > 3) {
    $compositeFilter = '(&('.$static_filter.')('.$interactive_filter.'))';
  } else {
    $compositeFilter = '('.$static_filter.')';
  }
  #print "LdapTreeView composite filter: $compositeFilter\n";

  my $result = $self->{ldap_source}->search(filter => $compositeFilter, base => $self->{base_dn}, attrs => ['dn']);
  my @entries = $result->sorted;
  my $tree_model = Gtk2::TreeStore->new(qw/Glib::String Glib::Boolean Glib::String/);
  my $prev_dn = undef;
  my $child; 
  my $entry;
  my @ancestors_stack = ();
  my %dn_iters;
  #push @parents, $toplevel;
  foreach $entry (@entries) {
    my $dn = Gtk2::Net::LDAP::Widgets::DistinguishedName->new($entry->dn);
    my $rdn;
    my $parent = $ancestors_stack[$#ancestors_stack];
    if (defined($prev_dn) && $dn->isDescendant($prev_dn)) {
      # it's a child of the previous dn
      # print "$dn is a child of $prev_dn\n";
      # TODO: assert length($entry->dn) =okolo (rindex($entry->dn, $prev_dn) + length($prev_dn))
      # TODO: push the previous DN onto a stack

      # Push the parent to stack:
      push(@ancestors_stack, $prev_dn);
      $parent = $ancestors_stack[$#ancestors_stack];
      
    } else {
      # it might not be a descendant of the parent anymore. Search for the youngest ancestor:
      while(scalar(@ancestors_stack)) {
        $parent = pop(@ancestors_stack);
        if ($dn->isDescendant($parent)) {
          push(@ancestors_stack, $parent);
          last;
        }
      }
      #$rdn = $dn->getRdn($parent);
    }
    # determine the RDN:
    $rdn = $dn->getRdn($parent);
    #print " ...so its RDN is $rdn\n";
    #print "number of components:".$dn->getLength."\n";

    # Determine the Iter-a of the super element in the tree model (if there's 
		# no parent, then iter is undefined and tree's top level is created):
    my $iter = undef;
    if (defined($parent)) {
      $iter = $dn_iters{$parent->{dn}};
    }

    $child = $tree_model->append($iter);

    $tree_model->set($child,
        0 => $rdn,
        1 => 0,
        2 => ($dn->{dntext})
        );
    $dn_iters{$dn->{dn}} = $child;
    $prev_dn = $dn;
  }

  $self->set_model($tree_model);

}
#---}}}

# by OLO
# czw mar 17 17:51:20 CET 2005
# Conversion of self to string:
sub to_string {
  my $self  = shift;
  return $self->{class}.' "'.\$self.'"';
}

#---[ sub get_dn ]---{{{

=head2 get_dn

Return the list of selected entries' Distinguished Names.

The list has at most one entry if single_selection is set to 1.

=cut
sub get_dn {
  my $self  = shift;
  my @dn_list;
  
  if ($self->{single_selection}) {
    push @dn_list, $self->{selectedDN};
  } else {
    my $model = $self->get_model;
    $model->foreach( sub {
      my $model = shift;
      my $path = shift;
      my $iter = shift;
      if ($model->get ($iter, $bool_column) > 0) {
        # The entry is selected:
        push(@dn_list, $model->get ($iter, $dn_column));
      }
      return 0;
    });
  }
  
  #print "Selected:\n";
  #print Dumper(\@dn_list);
  #print "\n";
  return @dn_list;
}
#---}}}

#---[ sub set_dn ]---{{{

=head2 set_dn

=over 4

=item set_dn( dn_list )

Sets the state of entries specified by DNs in dn_list to "selected" and unselects all other entries.

C<dn_list> list of Distinguished Names of entries to select

=back

=cut
sub set_dn(@) {
  my $self  = shift;
  my @dn_list = @_;

  # Build a hash map to speed up lookups:
  my %dn_hash = map { my $DN = new Gtk2::Net::LDAP::Widgets::DistinguishedName($_); ($DN->{dn}) => 1; } @dn_list;

  my $model = $self->get_model;
  $model->foreach( sub {
    my $model = shift;
    my $path = shift;
    my $iter = shift;
    my $row_dn = $model->get ($iter, $dn_column);
    my $row_DistinguishedName = new Gtk2::Net::LDAP::Widgets::DistinguishedName($row_dn);
    if ($dn_hash{$row_DistinguishedName->{dn}}) {
      # DN is on the supplied list, select the row:
      $model->set ($iter, $bool_column, 1);
    } else {
      # DN is not on the supplied list, deselect the row:
      $model->set ($iter, $bool_column, 0);
    }
    return 0;
  });
}
#---}}}

sub LdapTreeView_item_toggled {
  #my $self  = shift;
  my ($cell, $path_str, $self) = @_;
  my $model = $self->get_model;
  my $path = Gtk2::TreePath->new_from_string ($path_str);
  #print "$path_str\n$path\n";
  my $column = $cell->get_data ("column");
  # get toggled iter
  my $iter = $model->get_iter ($path);
  #print Dumper($iter);
  my ($toggle_item) = $model->get ($iter, $column);

  # do something with the value
  my $selectedDN = $model->get($iter, $dn_column);
  #print "Row: ".$model->get($iter, 0)."\n";
  #print "DN: ".$model->get($iter, 2)."\n";
  #print "toggle_item before: $toggle_item\n";

  if ($self->{single_selection}) {
    $model->foreach( sub { my $model = shift; my $path = shift; my $iter = shift; $model->set ($iter, $column, 0); return 0; } );
    $model->set ($iter, $column, 1);
    $self->{selectedDN} = $selectedDN;
  } else {
    $toggle_item ^= 1;
    $model->set ($iter, $column, $toggle_item);
  }

}

#---[ sub set_interactive_filter ]---{{{

=head2 set_interactive_filter

=over 4

=item set_interactive_filter ( interactive_filter )

Sets the interactive filter which is an additional filter applied to LDAP 
searches, usually provided by interactive components like a search/filter box 
and refreshes the data model, re-executing LDAP search and building a new data 
tree. 

C<interactive_filter> a string representation of an LDAP filter

=back

=cut
sub set_interactive_filter($) {
  my $self  = shift;
  $self->{interactive_filter} = shift;
  $self->refresh_model;
}
#---}}}

1;
__END__

#---[ pod end ]---{{{

=head1 SEE ALSO

L<Gtk2::Net::LDAP::Widgets>
L<Gtk2>
L<Net::LDAP>

=head1 AUTHOR

Aleksander Adamowski, E<lt>aleksander.adamowski@gmail.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright 2005,2008 by Aleksander Adamowski

This library is free software; you can redistribute it and/or modify it under
the same terms as Perl itself. 

=cut

#---}}}



Widgets/LdapTreeSelector.pm
# by OLO
# czw mar 17 14:10:53 CET 2005
# LdapTreeSelector: wielokrotna selekca elementow z drzewa LDAP

package Gtk2::Net::LDAP::Widgets::LdapTreeSelector;
#---[ pod head ]---{{{

=head1 NAME

Gtk2::Net::LDAP::Widgets::LdapTreeSelector - LDAP entry tree selection window

=head1 SYNOPSIS

    use Gtk2::Net::LDAP::Widgets;

    my $entryPopup = Gtk2::Net::LDAP::Widgets::LdapTreeSelector->new ($parent_window,
      $ldap_source,
      'o=People,dc=example,dc=com',
      '(|(objectClass=inetorgperson)(objectClass=organization))',
      'interactive_filter_type' => 'none',
      'single_selection' => 1
    );
    # expand entries below top-level tree entry:
    $entryPopup->{ldapTreeView}->expand_row(Gtk2::TreePath->new_from_string('0'), 0);
    $entryPopup->signal_connect (response => sub {
      my ($popup, $response) = @_;
      if($response =~ 'accept') {
        print "Selected entry DN: ".$entryPopup->get_dn;
      }
      $_[0]->destroy;
    });
    $entryPopup->show_all;

=head1 ABSTRACT

Gtk2::Net::LDAP::Widgets::LdapTreeSelector is a child class to L<Gtk2::Dialog> 
and is used to create a Gtk2 dialog which lets the user select LDAP entry/entries
displayed in a tree-like structure.

The dialog may be equipped with a search/filter box, although there might be 
problems with displaying search results when an interactive filter is set since 
there may be problems building the tree if all ancestors for an entry aren't 
included in the search result.

So it's advised to use this component without displaying a search/filter box 
('interactive_filter_type' => 'none').

=cut

#---}}}
use utf8;
use strict;
use vars qw(@ISA $VERSION);

use Net::LDAP;
use Net::LDAP::Util;
use Gtk2 -init;
use Data::Dumper;
use Gtk2::Net::LDAP::Widgets::DistinguishedName;

@ISA = qw(Gtk2::Dialog);

our $VERSION = "1.9.9";


use overload
q{""} => 'to_string';

# determine the filter (internal utility method)
sub _get_filter {
  my $self  = shift;
  if ($self->{interactive_filter_type} eq 'ldap') {
    return($self->{entryInteractiveFilter}->get_text);
  } elsif ($self->{interactive_filter_type} eq 'none') {
    return('');
  } elsif ($self->{interactive_filter_type} eq 'simple') {
    return('cn=*'.$self->{entryInteractiveFilter}->get_text.'*');
  }
  return('');
}

#---[ sub new ]---{{{

=head1 CONSTRUCTOR

=over 4

=item new ( parent, ldap_source, base_dn, static_filter, named parameters )

Creates a new Gtk2::Net::LDAP::Widgets::LdapTreeSelector object.

C<parent> the L<Gtk2::Window> which will be parent of this L<Gtk2::Dialog>

C<ldap_source> the L<Net::LDAP> object which is an active connection to an LDAP server

C<base_dn> the base DN of LDAP search operations

C<static_filter> the static filter that will be logically AND-ed with all filters executed by this selector

=back

=head2 named parameters

=over 4

=item init_interactive_filter =E<gt> 'some ldap filter'

The string to be initially put in the filter box

=item single_selection =E<gt> 0 | 1

Whether to use single selection mode (otherwise multiple selection is posible)


=item interactive_filter_type =E<gt> 'ldap' | 'simple' | 'none'

The type of filter box: 'ldap' supports full LDAP filter syntax, 'simple' does a substring search against the "cn" attribute, 'none' disables the search/filter box.

=back

=cut
sub new {
  my $class = shift;
  my $self = $class->SUPER::new('Wybierz wpis(y) LDAP', shift, 'destroy-with-parent',
                              'gtk-ok' => 'accept', 'gtk-cancel' => 'reject');
  $self->set_modal(1);
  $self->{ldap_source} = shift;
  $self->{base_dn} = shift;
  $self->{static_filter} = shift;

  my %named_params = @_;
  $self->{init_interactive_filter} = $named_params{'init_interactive_filter'};
  $self->{single_selection} = $named_params{'single_selection'};
  $self->{interactive_filter_type} = $named_params{'interactive_filter_type'};
  # possible values: 'ldap', 'simple', 'none':
  if (! $self->{interactive_filter_type}) {
    $self->{interactive_filter_type} = 'ldap';
  }

  my $btnFiltruj = Gtk2::Button->new_with_mnemonic ('_Filtruj');
  if ($self->{interactive_filter_type} ne 'none') {
    # The filter horizontal box:
    my $hboxFilter = Gtk2::HBox->new;
    my $labelFilter = Gtk2::Label->new;
    if ($self->{interactive_filter_type} eq 'ldap') {
      $labelFilter->set_markup("Wpisz <b>filtr LDAP</b>:");
    } elsif ($self->{interactive_filter_type} eq 'simple') {
      $labelFilter->set_markup("Wyszukaj:");
    } else {
      $labelFilter->set_markup("Wpisz <b>filtr</b>:");
    }
    $hboxFilter->pack_start ($labelFilter, 0, 0, 5);
    my $entryInteractiveFilter = Gtk2::Entry->new;
    $entryInteractiveFilter->set_text($self->{init_interactive_filter});
    $hboxFilter->pack_start ($entryInteractiveFilter, 1, 1, 5);

    $hboxFilter->pack_start ($btnFiltruj, 1, 1, 5);

    $self->{entryInteractiveFilter} = $entryInteractiveFilter;
    $self->vbox->pack_start ($hboxFilter, 0, 0, 5);
  }

  # Results list component:
  bless $self, $class;
  $self->{ldapTreeView} = Gtk2::Net::LDAP::Widgets::LdapTreeView->new($self->{ldap_source}, $self->{base_dn}, $self->{static_filter},
          $self->_get_filter, $self->{single_selection});
  
  my $scrollwin = Gtk2::ScrolledWindow->new;
  $scrollwin->set_policy ('never', 'automatic');
  $scrollwin->set_shadow_type ('in');
  $scrollwin->add($self->{ldapTreeView});
  $self->vbox->pack_start ($scrollwin, 1, 1, 5);

  $self->set_default_size(640, 480);

  $btnFiltruj->signal_connect (clicked => sub { 
    $self->refresh_list;
  });

  bless $self, $class;

}
#---}}}

# by OLO
# czw mar 17 17:51:20 CET 2005
# Conversion of self to string:
sub to_string {
  my $self  = shift;
  return $self->{class}.' "'.\$self.'"';
}

#---[ sub refresh_list ]---{{{
=head2 refresh_list

Refresh the entries list - re-execute the search with the filter determined by 
the search/filter box.

=cut
sub refresh_list {
  my $self = shift;
  my $newfilter = $self->_get_filter;
  #print "LdapTreeSelector nowy filtr: $newfilter\n";
  $self->{listEntriesView}->set_interactive_filter($newfilter);
}
#---}}}

#---[ sub get_dn ]---{{{

=head2 get_dn

Return the list of selected entries' Distinguished Names.

The list has at most one entry if single_selection is set to 1.

=cut
sub get_dn {
  my $self  = shift;
  return $self->{ldapTreeView}->get_dn;
}
#---}}}

1;

__END__

#---[ pod end ]---{{{

=head1 SEE ALSO

L<Gtk2::Net::LDAP::Widgets>
L<Gtk2>
L<Net::LDAP>

=head1 AUTHOR

Aleksander Adamowski, E<lt>aleksander.adamowski@gmail.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright 2005,2008 by Aleksander Adamowski

This library is free software; you can redistribute it and/or modify it under
the same terms as Perl itself. 

=cut

#---}}}


Widgets/DistinguishedName.pm
package Gtk2::Net::LDAP::Widgets::DistinguishedName;
#---[ pod head ]---{{{

=head1 NAME

Gtk2::Net::LDAP::Widgets::DistinguishedName - helper class for DN processing

=head1 SYNOPSIS

    This class is mostly used by other components to analyze and process LDAP 
    Distinguished Names and isn't meant to be used directly. Read the source in 
    case of any needs to do that.

=cut

#---}}}
use utf8;
use strict;
use vars qw(@ISA $VERSION);

use Net::LDAP;
use Net::LDAP::Util qw(ldap_explode_dn);
use Data::Dumper;

#@ISA = qw(Gtk2::TreeView);

our $VERSION = "1.9.9";

our $dntext;
our $dn;

use overload
q{""} => 'to_string';

# by OLO
# czw mar 17 17:51:34 CET 2005
# Constructor:
sub new {
  my $class = shift;
  my $self = {};
  $self->{dntext} = shift;
  $self->{dn} = _dn_normalize($self->{dntext});

  bless $self, $class;
}

# by OLO
# czw mar 17 17:51:20 CET 2005
# Conversion of self to string:
sub to_string {
  my $self  = shift;
  return $self->{dntext};
}

#---[ sub isDescendant ]---{{{

=head2 isDescendant

=over 4

=item isDescendant ( possible_ancestor )

Checks whether this DN is a child of another DN (possible_ancestor).

C<possible_ancestor> a distinguished name being tested for ancestorness in 
relation to current DN :)

=back

=cut
sub isDescendant($) {
  my $self  = shift;
  my $possible_ancestor = shift;
  my $mydn = $self->{dn};
  my $hisdn = $possible_ancestor->{dn};

  #print "Testing if $mydn is a child of $hisdn...\n";
  if (rindex($mydn, $hisdn) > 0) {
    #print "...it is.\n";
    return 1;
  } else {
    #print "...it isn't.\n";
    return 0;
  }
}
#---}}}

#---[ sub getRdn ]---{{{

=head2 getRdn

=over 4

=item getRdn ( possible_ancestor )

Returns the RDN of this DN in relation to a potential ancestor's DN ( the C<possible_ancestor> argument).
If the C<possible_ancestor> isn't an ancestor in fact, the returned RDN will be empty.

C<possible_ancestor> a distinguished name of potential ancestor

=back

=cut
sub getRdn($) {
  my $self  = shift;
  my $possible_ancestor = shift;
  if (! defined($possible_ancestor)) {
    $possible_ancestor = Gtk2::Net::LDAP::Widgets::DistinguishedName->new('');
  }
  my $mydn = $self->{dn};
  my $hisdn = $possible_ancestor->{dn};
  my $rdn = '';

  if (rindex($mydn, $hisdn) > 0) {
    # we're a descendant
    #print "$mydn is a descendant of $hisdn\n";

    # compute the RDN:
    $rdn = substr($mydn, 0, rindex($mydn, $hisdn));
    #print " ...so its RDN is $rdn\n";
  }
  $rdn =~ s/,+$//;
  $rdn =~ s/^,+//;
  return $rdn;
}
#---}}}

#---[ sub getLength ]---{{{

=head2 getLength

=over 4

=item getLength ( )

Returns the count of RDN components that this DN consists of.

=back

=cut
sub getLength {
  my $self  = shift;
  my @exploded_dn = @{ ldap_explode_dn($self->{dn}) };
  return scalar(@exploded_dn);
}
#---}}}

#---[ sub compareTo ]---{{{

=head2 compareTo

=over 4

=item compareTo ( )

Compares this DN and another DN using normalized forms of both. Return values
meanings are the same as with the "eq" Perl operator.

=back

=cut
sub compareTo($) {
  my $self = shift;
  my $him = shift;
  return ($self->{dn} eq $him->{dn});
}
#---}}}


# by OLO
# czw kwi 14 09:46:05 CEST 2005
# internal function used to normalize DNs for proper DN comparisons
sub _dn_normalize {
  my $dn = shift;
  $dn = lc($dn);
  $dn =~ s/\s//g;
  $dn = Net::LDAP::Util::canonical_dn($dn, casefold => 'lower');
  return $dn;
}

1;
__END__

#---[ pod end ]---{{{

=head1 SEE ALSO

L<Gtk2::Net::LDAP::Widgets>
L<Gtk2>
L<Net::LDAP>

=head1 AUTHOR

Aleksander Adamowski, E<lt>aleksander.adamowski@gmail.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright 2005,2008 by Aleksander Adamowski

This library is free software; you can redistribute it and/or modify it under
the same terms as Perl itself. 

=cut

#---}}}



Widgets/LdapEntryView.pm
package Gtk2::Net::LDAP::Widgets::LdapEntryView;
#---[ pod head ]---{{{

=head1 NAME

Gtk2::Net::LDAP::Widgets::LdapEntryView - LDAP entry viewport

=head1 SYNOPSIS

    This component is mostly used by other components and isn't meant to be 
used directly. Read the source in case of any needs to do that.

=cut

#---}}}
use utf8;
use strict;
use vars qw(@ISA $VERSION);

use Carp qw(cluck);
use Net::LDAP;
use Net::LDAP::Util;
use Gtk2 -init;
use Data::Dumper;
use Gtk2::Net::LDAP::Widgets::Util;

@ISA = qw(Gtk2::TreeView);

our $VERSION = "1.9.9";
our $dn_column = 0;
our $bool_column = 1;


use overload
q{""} => 'to_string';

# by OLO
# czw mar 17 17:51:34 CET 2005
# Constructor:
sub new {
  my $class = shift;
  my $self = $class->SUPER::new;
  $self->{ldap_source} = shift;
  $self->{base_dn} = shift;
  $self->{static_filter} = shift;
  $self->{interactive_filter} = shift;
  $self->{single_selection} = shift;

  $self->{selectedDN} = undef;

  ###
  # 1st column
  my $renderer = Gtk2::CellRendererText->new;
  my $col_offset = $self->insert_column_with_attributes
    (-1, "DN", $renderer,
     text => 0);
  my $column = $self->get_column ($col_offset - 1);
  $column->set_clickable (1);

  ###
  # 2nd column
  $renderer = Gtk2::CellRendererToggle->new;
  $renderer->set (xalign => 0.0);
  $renderer->set_data (column => 1);
  if ($self->{single_selection}) {
    $renderer->set_radio (1);
  }
  $renderer->signal_connect (toggled => \&LdapEntryView_item_toggled, $self);
  $col_offset = $self->insert_column_with_attributes
    (-1, "Y/N", $renderer,
     active => 1
    );
  $column = $self->get_column ($col_offset - 1);
  $column->set_clickable (1);

  bless $self, $class;
  $self->refresh_model;

  bless $self, $class;
}

# by OLO
# czw mar 17 17:51:20 CET 2005
# Conversion of self to string:
sub to_string {
  my $self  = shift;
  return $self->{class}.' "'.\$self.'"';
}

# by OLO
# wto kwi 19 12:15:31 CEST 2005
# Returns the list of selected entries' Distinguished Names.
sub get_dn {
  my $self  = shift;
  my @dn_list;
  if ($self->{single_selection}) {
    push @dn_list, $self->{selectedDN};
  } else {
    #TODO: multiple
    my $model = $self->get_model;
    $model->foreach( sub {
      my $model = shift;
      my $path = shift;
      my $iter = shift;
      if ($model->get ($iter, $bool_column) > 0) {
        # Wpis jest zaznaczony:
        push(@dn_list, $model->get ($iter, $dn_column));
      }
      return 0;
    });
  }
  
  #print "Selected:\n";
  #print Dumper(\@dn_list);
  #print "\n";
  return @dn_list;
}

sub LdapEntryView_item_toggled {
  my ($cell, $path_str, $self) = @_;
  my $model = $self->get_model;
  my $path = Gtk2::TreePath->new_from_string ($path_str);
  my $column = $cell->get_data ("column");
  # get toggled iter
  my $iter = $model->get_iter ($path);
  #print Dumper($iter);
  my ($toggle_item) = $model->get ($iter, $column);

  # do something with the value
  my $selectedDN = $model->get($iter, $dn_column);
  #print "val: ".$model->get($iter, $column)."\n";
  #print "toggle_item before: $toggle_item\n";
  #$toggle_item ^= 1;
  #print "toggle_item after: $toggle_item\n";
  #$model->set ($iter, $column, $toggle_item);

  # set new value
  if ($self->{single_selection}) {
    $model->foreach( sub { my $model = shift; my $path = shift; my $iter = shift; $model->set ($iter, $column, 0); return 0; } );
    $model->set ($iter, $column, 1);
    $self->{selectedDN} = $selectedDN;
  } else {
    $toggle_item ^= 1;
    $model->set ($iter, $column, $toggle_item);
  }

}

# by OLO
# wto kwi 19 13:19:38 CEST 2005
# Changes the interactive filter and refreshes data model
sub set_interactive_filter($) {
  my $self  = shift;
  $self->{interactive_filter} = shift;
  $self->refresh_model;
}

sub refresh_model {
  my $self  = shift;
  my $static_filter = $self->{static_filter};
  my $interactive_filter = $self->{interactive_filter};
  # Remove superfluous pairs of parentheses:
  $interactive_filter = filter_trim_outer_parens($interactive_filter);
  $static_filter = filter_trim_outer_parens($static_filter);

  my $compositeFilter;
  if (length($interactive_filter) > 3) {
    $compositeFilter = '(&('.$static_filter.')('.$interactive_filter.'))';
  } else {
    $compositeFilter = '('.$static_filter.')';
  }
  #print "LdapTreeView composite filter: $compositeFilter\n";

  my $result = $self->{ldap_source}->search(filter => $compositeFilter, base => $self->{base_dn}, attrs => ['dn']);
  my @entries = $result->sorted;
  my $newModel = Gtk2::ListStore->new (qw/Glib::String Glib::Boolean/);

  foreach my $entry (@entries) {
    #print "Entry: ".$entry->dn."\n";
    my $value = $newModel->set ($newModel->append, $dn_column => $entry->dn, $bool_column => 0);
  }
  $self->set_model($newModel);
}

1;

__END__

#---[ pod end ]---{{{

=head1 SEE ALSO

L<Gtk2::Net::LDAP::Widgets>
L<Gtk2>
L<Net::LDAP>

=head1 AUTHOR

Aleksander Adamowski, E<lt>aleksander.adamowski@gmail.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright 2005,2008 by Aleksander Adamowski

This library is free software; you can redistribute it and/or modify it under
the same terms as Perl itself. 

=cut

#---}}}
